### PR TITLE
storagetype API param in list service/disk offerings

### LINKED ISSUE #8206: Have a parameter "storagetype" for the listServiceOfferings and  listDiskOfferings api call
ISSUE TYPE

 Improvement request

**COMPONENT NAME**

Component: API

CLOUDSTACK VERSION

Cloudstack version 4.18

**SUMMARY**

Have a parameter "storagetype" for the following api calls

https://cloudstack.apache.org/api/apidocs-4.18/apis/listServiceOfferings.html
https://cloudstack.apache.org/api/apidocs-4.18/apis/listDiskOfferings.html

Currently, there is no way to list  the  service offering and disk  offerings based on the storage type (local and shared)

The create service offering , disk offering  there is a parameter called "storagetype"

 https://cloudstack.apache.org/api/apidocs-4.18/apis/createServiceOffering.html
 
 https://cloudstack.apache.org/api/apidocs-4.18/apis/createDiskOffering.html

Compute offering

Screenshot 2023-11-09 at 3 14 45 PM

Disk offering 
Screenshot 2023-11-09 at 3 15 02 PM

**It would be beneficial for the admin user to list only offerings based on storage type**

### PR DESCRIPTION
**Pull request by slavkap**
### Description

This PR fixes #8206 

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI

#### Feature/Enhancement Scale

- [ ] Major
- [X] Minor

### How Has This Been Tested?
```

(local) 🐵 > list serviceofferings storagetype=local
{
  "count": 1,
  "serviceoffering": [
    {
      "cpunumber": 1,
      "cpuspeed": 1000,
      "created": "2023-11-10T11:41:02+0200",
      "defaultuse": false,
      "diskofferingstrictness": false,
      "displaytext": "Medium Instance",
      "dynamicscalingenabled": true,
      "encryptroot": false,
      "hasannotations": false,
      "id": "045f2740-1570-481c-9a65-cc2ce3a44c63",
      "iscustomized": false,
      "issystem": false,
      "isvolatile": false,
      "limitcpuuse": false,
      "memory": 1024,
      "name": "Medium Instance",
      "offerha": false,
      "provisioningtype": "thin",
      "rootdisksize": 0,
      "storagetype": "local"
    }
  ]
}
(local) 🐵 > list serviceofferings storagetype=shared
{
  "count": 1,
  "serviceoffering": [
    {
      "cpunumber": 1,
      "cpuspeed": 500,
      "created": "2023-11-10T11:41:02+0200",
      "defaultuse": false,
      "diskofferingstrictness": false,
      "displaytext": "Small Instance",
      "dynamicscalingenabled": true,
      "encryptroot": false,
      "hasannotations": false,
      "id": "3ac4dd4f-6f2a-4d45-a0f7-45da41ae2abd",
      "iscustomized": false,
      "issystem": false,
      "isvolatile": false,
      "limitcpuuse": false,
      "memory": 512,
      "name": "Small Instance",
      "offerha": false,
      "provisioningtype": "thin",
      "rootdisksize": 0,
      "storagetype": "shared"
    }
  ]
}
(local) 🐵 > list diskofferings storagetype=Local
{
  "count": 1,
  "diskoffering": [
    {
      "created": "2023-11-10T11:41:02+0200",
      "disksize": 0,
      "disksizestrictness": false,
      "displayoffering": true,
      "displaytext": "Custom Disk",
      "encrypt": false,
      "hasannotations": false,
      "id": "87252a0c-a3ec-49f7-92fe-64660858e309",
      "iscustomized": true,
      "name": "Custom",
      "provisioningtype": "thin",
      "storagetype": "local"
    }
  ]
}
(local) 🐵 > list diskofferings storagetype=shared
{
  "count": 3,
  "diskoffering": [
    {
      "created": "2023-11-10T11:41:02+0200",
      "disksize": 5,
      "disksizestrictness": false,
      "displayoffering": true,
      "displaytext": "Small Disk, 5 GB",
      "encrypt": false,
      "hasannotations": false,
      "id": "cc62aed2-83e1-4871-8985-211ed08ab7b2",
      "iscustomized": false,
      "name": "Small",
      "provisioningtype": "thin",
      "storagetype": "shared"
    },
    {
      "created": "2023-11-10T11:41:02+0200",
      "disksize": 20,
      "disksizestrictness": false,
      "displayoffering": true,
      "displaytext": "Medium Disk, 20 GB",
      "encrypt": false,
      "hasannotations": false,
      "id": "a96419bd-3826-4f67-9295-96da19d64a29",
      "iscustomized": false,
      "name": "Medium",
      "provisioningtype": "thin",
      "storagetype": "shared"
    },
    {
      "created": "2023-11-10T11:41:02+0200",
      "disksize": 100,
      "disksizestrictness": false,
      "displayoffering": true,
      "displaytext": "Large Disk, 100 GB",
      "encrypt": false,
      "hasannotations": false,
      "id": "61ffdca0-081a-4208-9d22-1fa4c26793ca",
      "iscustomized": false,
      "name": "Large",
      "provisioningtype": "thin",
      "storagetype": "shared"
    }
  ]
}
```

### CODE REVIEW

File: server/src/main/java/com/cloud/api/query/QueryManagerImpl.java [hunk 1]
```diff
@@ -3181,6 +3184,13 @@ private Pair<List<DiskOfferingJoinVO>, Integer> searchForDiskOfferingsInternal(L
             sc.addAnd("encrypt", SearchCriteria.Op.EQ, encrypt);
         }
 
+        if (storageType != null) {
+            Boolean isLocalStorageType = getStorageType(storageType);
+            if (isLocalStorageType != null) {
+                sc.addAnd("useLocalStorage", Op.EQ, isLocalStorageType);
+            }
+        }
+
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/api/query/QueryManagerImpl.java:3193]:
```
        if (storageType != null) {
            Boolean isLocalStorageType = getStorageType(storageType);
            if (isLocalStorageType != null) {
                sc.addAnd("useLocalStorage", Op.EQ, isLocalStorageType);
            }
        }
```
this is added twice, can you extrract it to a method?

Comment (slavkap) [reply server/src/main/java/com/cloud/api/query/QueryManagerImpl.java:3193]: I think the code will be more complex if I try to get the class type of the SerachCriteria (sc) generic parameter.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by kiranchavala** [comment]
@blueorangutan package

**Review by kiranchavala — APPROVED** [review]
LGTM, tested manually 

```
CMK) > list serviceofferings storagetype=local
{
  "count": 1,
  "serviceoffering": [
    {
      "cacheMode": "none",
      "cpunumber": 2,
      "cpuspeed": 1500,
      "created": "2023-11-13T04:03:14+0000",
      "defaultuse": false,
      "diskofferingstrictness": false,
      "displaytext": "localservice",
      "dynamicscalingenabled": true,
      "encryptroot": false,
      "hasannotations": false,
      "id": "b700afee-3a8b-4f76-b705-cb0d0a74bfa8",
      "iscustomized": false,
      "issystem": false,
      "isvolatile": false,
      "limitcpuuse": false,
      "memory": 2048,
      "name": "localservice",
      "offerha": false,
      "provisioningtype": "thin",
      "rootdisksize": 0,
      "storagetype": "local"
    }
  ]
}
(CMK) > list diskofferings storagetype=local
{
  "count": 1,
  "diskoffering": [
    {
      "cacheMode": "none",
      "created": "2023-11-13T04:03:33+0000",
      "disksize": 0,
      "disksizestrictness": false,
      "displayoffering": true,
      "displaytext": "localdisk",
      "encrypt": false,
      "hasannotations": false,
      "id": "96108615-c1f3-41c9-93c7-8f03f3ef27af",
      "iscustomized": true,
      "name": "localdisk",
      "provisioningtype": "thin",
      "storagetype": "local"
    }
  ]
}
```

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Review by DaanHoogland — APPROVED** [review]
clgtm

**Comment by DaanHoogland** [comment]
@blueorangutan test

File: server/src/main/java/com/cloud/api/query/QueryManagerImpl.java [hunk 2]
```diff
@@ -3541,6 +3560,7 @@ private Pair<List<ServiceOfferingJoinVO>, Integer> searchForServiceOfferingsInte
         return _srvOfferingJoinDao.searchAndCount(sc, searchFilter);
     }
 
+
```
Comment (harikrishna-patnala) [inline server/src/main/java/com/cloud/api/query/QueryManagerImpl.java:3563]: minor one, see if this can be removed

**Review by harikrishna-patnala — APPROVED** [review]
code LGTM

File: server/src/main/java/com/cloud/api/query/QueryManagerImpl.java (hunk 2 above)
Comment (DaanHoogland) [reply server/src/main/java/com/cloud/api/query/QueryManagerImpl.java:3563]:
```suggestion
```

**Pull request merged**